### PR TITLE
Added target and motion events

### PR DESCRIPTION
**Pull request by couger01**
* Updated simulator to use simactuators
* Added motion and target events
* Improved unit test coverage
* Prepared for salobj v6

### CODE REVIEW

File: python/lsst/ts/cbp/component.py [hunk 1]
```diff
@@ -192,6 +229,8 @@ async def move_azimuth(self, position: float):
         if position < -45 or position > 45:
             raise ValueError("New azimuth value exceeds Azimuth limit.")
         else:
+            self.azimuth_target = position
+            self.in_motion = True
```
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/cbp/component.py:233]: Also consider caching some of this data in the event you use to report it.

Advantages include:
* Changes are easily tracked and reported.
* There is no danger of one cache (in the model) going out of date with another (in the CSC).
* No need to call back to the CSC every time the internal state changes.
* Such callbacks are likely expensive because the CSC has to assume anything might have changed, whereas the model knows exactly what has changed.
Disadvantages:
* The modularity suffers. The SAL events leak in to the model. I often handle this by making the model contain a copy of the CSC, so it can get to the events. It's circular, but it works.

For example if your CSC has an axesInPostion event then you could
```
self.csc.evt_axesInPosition.set_put(azimuth=False, force_output=True)
```
To ask for the most recently reported value (which is not necessarily the same as the actual current value, but should be exactly the same as any other cached value, such as from self.encoder_motion):
```
azimuth_in_position = self.csc.evt.axesInPosition.data.azimuth
```
And of course you have to poll the actuators or something to keep the data up to date. The polling loop would set:
```
azimuth_in_position = ... as above
elevation_in_position = ...
self.csc.evt_axesInPosition.set_put(
    azimuth=azimuth_in_position,
    elevation=elevation_in_position,
    ... # do NOT end with force_output=True
)
```

File: python/lsst/ts/cbp/component.py [hunk 2]
```diff
@@ -94,28 +102,55 @@ def __init__(self):
...
         mask_dict = {
             f"{i}": types.SimpleNamespace(name=f"Empty {i}", rotation=0, id=i)
             for i in (1, 2, 3, 4, 5, 9)
         }
         mask_dict["9"].name = "Unknown"
         self.masks = types.SimpleNamespace(**mask_dict)
-        self.log.info(f"{self.masks}")
+
+    def get_motion_status(self):
+        self.encoder_motion.AZIMUTH = self.azimuth != self.azimuth_target
```
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/cbp/component.py:136]: On the other hand, if you are going to report this value as a SAL event, then I suggest you use the event to cache the information. You already have it in the CSC; you'd have to feed it to the model. That avoids needlessly duplicating cached data.

If your CSC has an axesInPosition event that you pass to the model, then whenever you start a move you set:
```
self.evt_axesInPosition.set_put(azimuth=False)
```

If code wants the value that was most recently reported it can read the cache using:
```
azimuth_in_position = self.evt_axesInPosition.data.azimuth
```

And in the code that polls the actuators:
```
azimuth_in_position = ... read azimuth encoder and "is moving" and compute the value, as above
...likewise for the other axes...
self.evt_axesInPosition.set_put(
    azimuth=azimuth_in_position,
    elevation=elevation_in_position,
    ....
)
```
You already should be reporting this, and it avoids a mismatch between your model and what your CSC reports.

File: python/lsst/ts/cbp/component.py [hunk 3]
```diff
@@ -362,11 +412,11 @@ async def check_cbp_status(self):
         None
 
         """
-        self.encoder_status.AZIMUTH = bool(await self.send_command("AAstat=?"))
-        self.encoder_status.ELEVATION = bool(await self.send_command("ABstat=?"))
-        self.encoder_status.MASK_SELECT = bool(await self.send_command("ACstat=?"))
-        self.encoder_status.MASK_ROTATE = bool(await self.send_command("ADstat=?"))
-        self.encoder_status.FOCUS = bool(await self.send_command("AEstat=?"))
+        self.encoder_status.AZIMUTH = bool(int(await self.send_command("AAstat=?")))
```
Comment (r-owen) [inline python/lsst/ts/cbp/component.py:415]: Surely this should be the motion fields? Else why cast to boolean?

**Review by r-owen — COMMENTED** [review]
I am uneasy about the way data is cached and I'm not sure I understand exactly what the Status and Motion classes (I thought the former was encoder position, but later code sets the fields to bool).

File: python/lsst/ts/cbp/component.py [hunk 4]
```diff
@@ -3,18 +3,26 @@
...
 import asyncio
 import types
-import enum
 
 
-class Status(enum.Flag):
-    """The status for the encoders.
-    """
+class Status:
+    """The status for the encoders."""
```
[the remark this answers is not in the document]
Comment (couger01) [reply python/lsst/ts/cbp/component.py:9]: Actually its the encoder status, which is basically if the encoder have a problem or not.

File: python/lsst/ts/cbp/component.py (hunk 3 above)
Comment (couger01) [reply python/lsst/ts/cbp/component.py:415]: Its the status of the encoder.

File: python/lsst/ts/cbp/component.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (couger01) [reply python/lsst/ts/cbp/component.py:136]: 1. No.
2. Added error tolerance
Improved in_position definition based on suggestions.

File: python/lsst/ts/cbp/component.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/cbp/component.py:9]: Please explain that in the doc string, including what the values mean.

Please consider a less generic name, such as `EncoderStatus`.

If the values are booleans then please initialize them to False, rather than 0.

File: python/lsst/ts/cbp/component.py [hunk 5]
```diff
@@ -3,18 +3,26 @@
...
+
+    def __init__(self):
+        self.AZIMUTH = 0
```
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/cbp/component.py:12]: I suggest waiting to resolve the conversation until you make the change. It makes it easier to find.

File: python/lsst/ts/cbp/component.py (hunk 3 above)
Comment (r-owen) [reply python/lsst/ts/cbp/component.py:415]: Got it. So once the doc string for the class is updated and the values initialized to False instead of 0, this should be be clear.

**Review by r-owen — COMMENTED** [review]
Thanks for agreeing to make some of those changes. A few responses that might clarify.

File: python/lsst/ts/cbp/component.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (couger01) [reply python/lsst/ts/cbp/component.py:233]: Added some more subtests for testing if moving to the same position. Updated the CSC to handle this case. Also changed moving to in_position as that more accurately reflects what the information received.

File: python/lsst/ts/cbp/component.py [hunk 6]
```diff
@@ -192,7 +237,10 @@ async def move_azimuth(self, position: float):
         if position < -45 or position > 45:
             raise ValueError("New azimuth value exceeds Azimuth limit.")
         else:
-            await self.send_command(f"new_az={position}")
+            if self.azimuth_target != position:
```
Comment (r-owen) [inline python/lsst/ts/cbp/component.py:240]: I fear this test may not be sufficient. What if the target matches but the actuator didn't actually arrive at the target? Surely the user should be able to try the move again?

If so, I see two reasonable options:
* Eliminate the test. If moving is simple and safe then I suggest that.
* If moves are expensive or risky, then test if the target matches AND the current position is not close enough to that position. I suspect for that test to be robust you will have to explicitly ask for the current position instead of relying on the telemetry loop to have done that (sufficiently recently) for you.

This comment applies to all actuator move methods.

File: python/lsst/ts/cbp/csc.py [hunk 7]
```diff
@@ -83,14 +82,32 @@ async def do_move(self, data):
...
         self.assert_enabled("move")
-        await asyncio.join(
-            [
-                self.model.move_elevation(data.elevation),
-                self.model.move_azimuth(data.azimuth),
-            ]
+        await asyncio.gather(
+            self.model.move_elevation(data.elevation),
+            self.model.move_azimuth(data.azimuth),
+        )
```
Comment (r-owen) [inline python/lsst/ts/cbp/csc.py:88]: Does this only move in elevation and azimuth?

If so, please refactor the code below so that it can also be called by the methods that trigger motion of the other actuators.

Even if it moves all axes, please consider having a generic "report_state" method that you can call (preferably from a callback function in the Model whenever state changes).

File: python/lsst/ts/cbp/csc.py [hunk 8]
```diff
@@ -83,14 +82,32 @@ async def do_move(self, data):
...
+            mask=self.model.encoder_motion.mask_select,
+            mask_rotation=self.model.encoder_motion.mask_rotate,
+        )
+        await asyncio.wait_for(self.in_position(), 20)
+        self.evt_inPosition.set_put(
+            azimuth=self.model.encoder_motion.azimuth,
+            elevation=self.model.encoder_motion.elevation,
+            focus=self.model.encoder_motion.focus,
+            mask=self.model.encoder_motion.mask_select,
+            mask_rotation=self.model.encoder_motion.mask_rotate,
```
Comment (r-owen) [inline python/lsst/ts/cbp/csc.py:109]: I very strongly suggest centralizing your event-putting code and improving the robustness of the state reporting, to make sure the model state and reported state always match. Here are two techniques I have used that work:

* Pass the CSC to the model and have it use the CSC's event topics to store the state. That avoids duplication and simplifies state reporting. Instead of the CSC having to read all state from the model every time any state changes , the state is automatically reported when it changes. Concretely, in the model this means calling `self.csc.evt_foo.set_put(...)` to set state and `self.csc.evt_foo.data` to read state.

This is simple and robust, and minimizes duplication. You can use accessors to simplify getting and setting state, which adds some boilerplate code but makes the code more readable.

* Make the event reporting a separate method of your CSC and calling it *from the model* as a callback function, whenever the model updates any state.

This requires duplicating state between the model and the CSC, and requires that the CSC read all state any time any state changes.

**Review by r-owen — CHANGES_REQUESTED** [review]
Thank you for the improvements.

My main concerns at this point are:
* Simplifying and making more robust the reporting of state changes in the model. At present I'm not convinced that connection is robust, and it definitely has too much duplication of code (e.g. calling evt_foo.set_put in too many places).
* Changing the term "motion" to "in position" to reduce confusion.

File: python/lsst/ts/cbp/component.py (hunk 6 above)
Comment (couger01) [reply python/lsst/ts/cbp/component.py:240]: Eliminated the test.

**Review by r-owen — CHANGES_REQUESTED** [review]
I fear that your code is still not reliably outputting the events such as inPosition. If so I would really like to see this thoroughly fixed.

Consider the do_move command: I would expect some code to set *and report* inPosition=False when the move starts, then your telemetry loop will eventually report it true again. As it stands I am pretty sure you have a race condition: if the next poll occurs after the move finishes then the user will never see inPosition=False and may not realize that a move occurred.

Also, two minor changes to consider:
* Rename `publish` to `read_status` or `get_status` or similar, in `CBPComponent`, since it does not seem to publish anything (certainly not upward to the CSC or SAL).
* Make the model naming more consistent, e.g. by changing the CSC attribute from `model` to `component` or renaming the CBPComponent to CBPModel.

File: python/lsst/ts/cbp/component.py [hunk 9]
```diff
@@ -136,29 +143,63 @@ def generate_mask_info(self):
...
-    def get_in_position_status(self):
-        """Get the in position status of each actuator."""
-        self.encoder_in_position.azimuth = (
+    def get_in_position_status(self, changed=True, move_commanded=[]):
+        """Get the in position status of each actuator.
+
+        Also calls a method to publish the inPosition event if the status of
+        any actuator has changed since the previous call.
+        """
+        changed = changed
```
Comment (r-owen) [inline python/lsst/ts/cbp/component.py:152]: I suggest that you store in_position, target position and current position **only** in the SAL topics, instead of keeping a duplicate of this information in your own custom classes.

This eliminates a lot of custom code and also eliminates any risk of the two copies of the information getting out of synch with each other.

Here is what that looks like:

1) The AxisInPosition class and its cousins disappear.

2) Use accessors to get the data from the topics:
```
@property
def in_position(self):
    return self.csc.evt_inPosition.data

@property
def target(self):
    return self.csc.evt_target.data

@propery
def azimuth(self):
    return self.csc.tel_CBP_azimuth.azimuth

@property
def elevation(self):
    return self.csc.tel_CBP_elevation.elevation

... etc. for the other axes.
```

3) In the move commands, explicitly set the fields false for the axes being moved, once you know that the move has successfully started. For example:
```
def move_azimuth(self, azimuth):
    # ... do all your usual stuff here, and if the move command succeeds
    self.csc.evt_inPosition.set_put(azimuth=False)
```

4) This entire method (and its cousins that set your other custom classes) are much simpler; just set the data once and you're done. Here is the rewritten version of this method, which I also renamed because it doesn't get anything; it just sets. It is dramatically simpler and yet it does everything needed:
```
def update_in_position(self):
    """Update the in position status of each actuator,
    based on the most recently read encoder data.

    Returns
    --------
    did_change : `bool`
        True if anything changed (and so the event was published)
    """
    return self.csc.evt_inPosition.set_put(
        azimuth=abs(self.azimuth - self.target.azimuth) < self.error_tolerance,
        elevation=abs(self.elevation - self.target.altitude) < self.error_tolerance,
        mask=self.mask == self.target.mask,
        mask_rotation=(
            abs(self.mask_rotation - self.target.mask_rotation) < self.error_tolerance
        ),
        focus=abs(self.focus - self.target.focus) < self.error_tolerance,
    )
```
If your code ever needs to force this even to be published then you can add a `force_output` argument that defaults to False. `force_output` is a valid keyword argument for the `set_put` method of event topics (but not telemetry topics because telemetry set_put *always* outputs the topic).

Comment (r-owen) [reply python/lsst/ts/cbp/component.py:152]: I am concerned that your error tolerance is the same for all axes. Is that safe? I would expect it to work for azimuth and elevation (since the encoder resolution and motion accuracy is probably similar for both) but I am concerned about mask rotation and focus.

If you have one tolerance for each axis (with a value of 0 for mask) you could even automate the in_position code, which nicely avoid typos but is a bit harder to read (you need `<=` instead of `<` if a tolerance can be 0):
```
in_position_kwargs = {
    axis: abs(
        getattr(self, axis) - getattr(self.target, axis)
    )
    <= getattr(self.tolerance, axis)
    for axis in self.axis_names
}
self.csc.evt_inPosition.set_put(**in_position_kwargs)
```

**Review by r-owen — COMMENTED** [review]
I hope you will consider storing the data in the topics. I supplied a code example. I am happy to discuss this if you are interested and have any questions.

**Review by r-owen — CHANGES_REQUESTED** [review]
This is close. However, I found some things, especially in the mock server, that I hope you will address. The biggest that comes to mind is rejecting invalid parameters.

File: python/lsst/ts/cbp/component.py [hunk 10]
```diff
@@ -213,7 +213,9 @@ async def move_azimuth(self, position: float):
 
         """
         if position < -45 or position > 45:
-            raise ValueError("New azimuth value exceeds Azimuth limit.")
+            raise ValueError(
+                f"New azimuth value {position:0.2f} exceeds Azimuth limit [-45, 45]."
+            )
```
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/cbp/component.py:218]: Also, please consider a function that takes the value, the limits and a brief description of the value. That avoids the danger that the numerical value of a limit and the reported value can differ if somebody changes one and forgets to change the other, and simplifies the code a bit. e.g.:
```
def assert_in_range(name, value, min_value, max_value):
    """Raise ValueError if a value is out of range."""
    if value < min_value or value > max_value:
        raise ValueError(f"{name} = {value} not in range [{min_value}, {max_value}]")
```
(if you avoid the format then it can be used for integers as well).

File: python/lsst/ts/cbp/csc.py [hunk 11]
```diff
@@ -80,10 +83,15 @@ async def telemetry(self):
...
         while True:
-            self.log.debug("Begin sending telemetry")
-            await self.component.update_status()
-            if self.component.status.panic:
-                self.fault(1, "CBP Panicked. Check hardware and reset device.")
+            try:
+                self.log.debug("Begin sending telemetry")
+                await self.component.update_status()
+                if self.component.status.panic:
+                    self.fault(1, "CBP Panicked. Check hardware and reset device.")
```
Comment (r-owen) [inline python/lsst/ts/cbp/csc.py:90]: I gently suggest adding a `return` after `self.fault(...)` to make it clear that the telemetry loop will not continue. It's not essential because going into FAULT state kills the the telemetry loop (eventually) but it makes sure that the telemetry loop cannot possibly run again while transitioning to fault state.

File: python/lsst/ts/cbp/csc.py [hunk 12]
```diff
@@ -31,6 +31,8 @@ class CBPCSC(salobj.ConfigurableCsc):
     component : `CBPComponent`
     simulator : `None` or `MockServer`
     telemetry_task : `asyncio.Future`
+    telemetry_time : `float`
+    in_position_time : `int`
```
Comment (r-owen) [inline python/lsst/ts/cbp/csc.py:35]: Please document what this is, and consider renaming it to `in_position_timeout`.

Also (I should have mentioned this earlier) please consider renaming "telemetry_time" to "telemetry_interval".

"time" is a bit overloaded. I try to always name time limits "timeout" because that matches the usage in `asyncio` and try to reserve `time` for absolute times.

File: python/lsst/ts/cbp/mock_server.py [hunk 13]
```diff
@@ -70,49 +70,42 @@ def __init__(self, log=None):
...
+            (re.compile(r"autopark=\?"), self.do_autopark),
+            (re.compile(r"park=(?P<parameter>[\?01])"), self.do_park),
+            (re.compile(r"AAstat=\?"), self.do_aastat),
+            (re.compile(r"ABstat=\?"), self.do_abstat),
+            (re.compile(r"ACstat=\?"), self.do_acstat),
+            (re.compile(r"ADstat=\?"), self.do_adstat),
+            (re.compile(r"AEstat=\?"), self.do_aestat),
+        )
         self.log = logging.getLogger(__name__)
+        print("MockServer created")
```
Comment (r-owen) [inline python/lsst/ts/cbp/mock_server.py:108]: Please delete this. (I assume it was accidentally left in from debugging).

File: python/lsst/ts/cbp/mock_server.py [hunk 14]
```diff
@@ -136,43 +129,34 @@ async def cmd_loop(self, reader, writer):
...
-                        if parameter is None:
-                            msg = await called_command() + "\r"
-                        else:
-                            msg = await called_command(parameter) + "\r"
-                        msg = msg.encode("ascii")
-                        writer.write(msg)
-                        self.log.debug(f"Wrote {msg}")
-                        await writer.drain()
+                    self.log.debug(f"{line} match: {matched_command}")
+                    self.log.debug(f"method: {command_method}")
```
Comment (r-owen) [inline python/lsst/ts/cbp/mock_server.py:146]: I gently suggest combining these two log messages into one.

**Review by r-owen — CHANGES_REQUESTED** [review]
This looks great.

My one serious concern is the way you are setting the target event at startup. I hope you can fix that to avoid having 4-5 target events at startup, all but the last with incorrect values in one or more fields.

File: doc/conf.py [hunk 15]
```diff
@@ -5,11 +5,8 @@
...
-    )
-)
-
-intersphinx_mapping["ts_xml"] = ("https://ts-xml.lsst.io", None)
-intersphinx_mapping["ts_salobj"] = ("https://ts-salobj.lsst.io", None)
+_g.update(build_package_configs(project_name="ts_CBP", version=lsst.ts.cbp.__version__))
+
+intersphinx_mapping["ts_xml"] = ("https://ts-xml.lsst.io", None)  # noqa
+intersphinx_mapping["salobj"] = ("https://ts-salobj.lsst.io", None)  # noqa
+intersphinx_mapping["simactuators"] = ("https://ts-simactuators.lsst.io", None)  # noqa
```
Comment (r-owen) [inline doc/conf.py:12]: Why did you remove the ts_ prefix from salobj and omit it from simactuator (while leaving it for ts_xml)? It presumably works either way, but I'm curious. I have been using the package name as the key (ts_xml, ts_salobj...).

File: python/lsst/ts/cbp/mock_server.py [hunk 16]
```diff
@@ -85,65 +169,205 @@ async def cmd_loop(self, reader, writer):
...
+
+        Parameters
+        ----------
+        altitude : `float`
+
+        Returns
+        -------
+        str
+        """
+        self.encoders.ELEVATION.set_position(float(altitude))
```
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/cbp/mock_server.py:203]: I'll correct my previous statement. I strongly suggest making your mock controller handle out-of-range positions (at least roughly) the same way as the real low-level controller.

I fervently hope the real low-level controller rejects such commands. If it does not then please explain what it does in the doc string for your mock controller class and emulate that.

And whatever you do, please test it in your unit test of the mock controller!

It is important to handle things like this correctly because your CSC may well end up sending an invalid command at some point (you have logic that rejects invalid positions in the CSC, which is good, but what if some future update means you have a limit wrong or that somebody breaks that code and the limit is not properly applied?).

File: python/lsst/ts/cbp/csc.py [hunk 17]
```diff
@@ -86,14 +87,23 @@ async def telemetry(self):
...
                 self.log.debug("Begin sending telemetry")
                 await self.component.update_status()
+                if not self.evt_target.has_data:
+                    self.evt_target.set_put(
+                        azimuth=self.component.azimuth,
+                        elevation=self.component.elevation,
+                        focus=self.component.focus,
+                        mask=self.component.mask,
+                        mask_rotation=self.component.mask_rotation,
+                    )
```
Comment (r-owen) [inline python/lsst/ts/cbp/csc.py:97]: Thanks!

**Review by r-owen — CHANGES_REQUESTED** [review]
I think the only thing left is handling invalid commands in MockServer.

**Comment by r-owen** [comment]
After a slack chat we realized that the command loop in the mock controller appears to be unsafe: if the do_x method that is called raises an error it looks as if the command loop exits (and without acknowledging the command as failed).

**Review by r-owen — CHANGES_REQUESTED** [review]
I'm afraid the command loop still needs work.

File: python/lsst/ts/cbp/mock_server.py [hunk 18]
```diff
@@ -160,13 +160,20 @@ async def cmd_loop(self, reader, writer):
...
-                            raise asyncio.TimeoutError()
+                    except ValueError as e:
+                        self.log.info(f"command {line} failed: {e}")
+                        # TODO DM-27693: reply with an error signal so the
+                        # client knows there is a problem
+                    except Exception:
+                        # An unexpected error; log a traceback
+                        self.log.exception("Bug! Command {line} filed.")
+                        # TODO DM-27693: reply with an error signal so the
+                        # client knows there is a problem
```
Comment (r-owen) [inline python/lsst/ts/cbp/mock_server.py:176]: Almost right. Unfortunately you then write `msg` which is the value from the previous iteration of the loop.

There are many ways around that. Here are two I like:
* Put the write/debug/drain in an `else` clause. That only runs if no except is raised.
* Initialize msg to None and only write it if it is not None.

(One I beg you NOT to use is to put a `continue` in each `except` clause, to skip the write/debug/drain. That's far too risky, because forgetting the `continue` will cause very subtle bugs (especially if the first command fails -- which will lead to using an undefined variable!)).

**Review by r-owen — CHANGES_REQUESTED** [review]
The command loop is close but still needs a bit of work.

File: python/lsst/ts/cbp/component.py [hunk 19]
```diff
@@ -165,250 +186,233 @@ async def disconnect(self):
...
+        self.masks.get("1").name = config.mask1["name"]
+        self.masks.get("1").rotation = config.mask1["rotation"]
+        self.masks.get("2").name = config.mask2["name"]
+        self.masks.get("2").rotation = config.mask2["rotation"]
+        self.masks.get("3").name = config.mask3["name"]
+        self.masks.get("3").rotation = config.mask3["rotation"]
+        self.masks.get("4").name = config.mask4["name"]
+        self.masks.get("4").rotation = config.mask4["rotation"]
+        self.masks.get("5").name = config.mask5["name"]
+        self.masks.get("5").rotation = config.mask5["rotation"]
```
Comment (r-owen) [inline python/lsst/ts/cbp/component.py:385]: I suggest replacing `.get("n")` with `["n"]`. This improves the error reporting if the key is missing. The current code will complain that "None has no attribute name" and not tell you wish key is the problem. With `["n"]` Python will complain that key "n" is missing and not try to use None as a mask.

File: doc/conf.py (hunk 15 above)
Comment (r-owen) [reply doc/conf.py:12]: I'm still curious about this. It's an odd inconsistency.

**Review by r-owen — APPROVED** [review]
It looks very good. A nice improvement.

I have two small changes for your consideration and one question. All optional.
